docs/writing-rules/pattern-syntax.mdx

```diff
@@ -492,34 +492,23 @@ Creating a buffer using size
 
 Semgrep substitutes the $X with the value detected within the metavariable. In this case, the value was `size`.
 
-However, in some cases, the matched value of the metavariable is not the real value you were looking for. See the following formula and testing code:
+To display matched metavariable in rule message:
 
-Part of a Semgrep rule (formula):
+1. Find the metavariable used in Semgrep rule. For example:
+    ```yaml
+    - pattern: |
+      byte[] buf = new byte[$X];
+    ```
+    Uses `$X` as a metavariable.
+2. Insert the metavariable to rule message:
+    ```yaml
+    - message: >-
+      Creating a buffer using $X
+    ```
 
-```yaml
-- patterns:
-   - pattern: byte[] buf = new byte[$X];
-   - metavariable-comparison:
-        metavariable: $X
-        comparison: $X < 2048
-```
-
-Testing code:
-
-```java
-int size = 512;
-byte[] buf = new byte[size];
-```
-
-Semgrep matches this code because it performs constant propagation. Therefore, Semgrep recognizes that the value of `size` is `512`. Consequently, Semgrep evaluates that the buffer size is less than `2048`. But what is the value of `$X`?
-
-If the rule message states `Creating a buffer using $X`, the resulting message output is not helpful in this particular case: `Creating a buffer using size`. This is caused by the value of `$X` within the code, which is `size`. However, the underlying value of `size` is `512`. The goal of the rule message is to access this underlying value in our message.
-
-To retrieve the correct value in the case described above, use `value($X)` in the rule message (for example (`Creating a buffer using value($X)`). Semgrep replaces the `value($X)` with the underlying propagated value of the metavariable `$X` if it computes one (otherwise, Semgrep uses the matched value).
-
-Run the following example in Semgrep Playground to see the message (click **Open in Editor**, and then **Run**, unroll the **1 Match** to see the message):
-
-<iframe title="Metavariable value in message example" src="https://semgrep.dev/embed/editor?snippet=returntocorp:value-in-message-example" width="100%" height="432" frameborder="0"></iframe>
+:::info
+There are specific limitations for nested variables which Semgrep cannot process correctly under specific circumstances. For more information see documentation of experimental feature [Displaying propagated value of metavariable](/experiments/display-propagated-metavariable-message)
```

Comment:
Can we call them "nested variables"? @emjin

Reply:
No, that's not really right. Also I wouldn't really call this a "limitation". It makes it sound like sometimes we don't have a matched value. Maybe we can say "If you're using Semgrep's advanced dataflow features, you might want to access more information than is present now. etc."

Reply:
Ok, changing to: If you're using Semgrep's advanced dataflow features, see documentation of experimental feature [Displaying propagated value of metavariable](/experiments/display-propagated-metavariable-message).